include/deal.II/lac/utilities.h

```diff
@@ -144,6 +208,95 @@ namespace Utilities
 {
   namespace LinearAlgebra
   {
+
+    template<typename NumberType>
+    std::array<NumberType,3> hyperbolic_rotation(const NumberType &f,
+                                                 const NumberType &g)
+    {
+      Assert (f != 0, ExcDivideByZero());
+      const NumberType tau = g/f;
+      AssertThrow (std::abs(tau) < 1.,
+                   ExcMessage("real-valued Hyperbolic rotation does not exist for ("+
+                              std::to_string(f) +
+                              "," +
+                              std::to_string(g)+
+                              ")"));
+      const NumberType u = std::copysign(sqrt((1.-tau)*(1.+tau)), f); // <-- more stable than std::sqrt(1.-tau*tau)
+      std::array<NumberType,3> csr;
+      csr[0] = 1./u;          // c
+      csr[1] = csr[0] * tau;  // s
+      csr[2] = f *u;          // r
+      return csr;
+    }
+
+    template<typename NumberType>
+    std::array<std::complex<NumberType>,3> hyperbolic_rotation(const std::complex<NumberType> &f,
+                                                               const std::complex<NumberType> &g)
+    {
+      AssertThrow(false, ExcNotImplemented());
+      std::array<NumberType,3> res;
+      return res;
+    }
+
+
+    template<typename NumberType>
+    std::array<std::complex<NumberType>,3> Givens_rotation(const std::complex<NumberType> &f,
+                                                           const std::complex<NumberType> &g)
+    {
+      AssertThrow(false, ExcNotImplemented());
+      std::array<NumberType,3> res;
+      return res;
+    }
+
+    template<typename NumberType>
+    std::array<NumberType,3> Givens_rotation(const NumberType &f,
+                                             const NumberType &g)
+    {
+      std::array<NumberType,3> res;
+      // naieve calculation for "r" may overflow or underflow:
```

Comment:
Let us say `naive` here rather than trying to typeset the diaeresis with ascii characters